systems/plants/TimeSteppingRigidBodyManipulator.m

```diff
@@ -366,58 +366,70 @@ function checkDirty(obj)
           J_check = zeros(0,num_q);
           JL = zeros(0,num_q^2);
         end
-
-        if (nContactPairs > 0)
+        
+        has_contacts = (nContactPairs > 0);
+        
+        if has_contacts
           if (nargout>4)
             [phiC,normal,d,xA,xB,idxA,idxB,mu,n,D,dn,dD] = obj.manip.contactConstraints(kinsol, obj.multiple_contacts);
           else
             [phiC,normal,d,xA,xB,idxA,idxB,mu,n,D] = obj.manip.contactConstraints(kinsol, obj.multiple_contacts);
           end
-
-          if isempty(possible_contact_indices)
-            possible_contact_indices = (phiC+h*n*qd) < obj.z_inactive_guess_tol;
-          end
-
-          nC = sum(possible_contact_indices);
-          mC = length(D);
-
-          J = zeros(nL + nP + (mC+2)*nC,num_q)*q(1); % *q(1) is for taylorvar
-          lb = zeros(nL+nP+(mC+2)*nC,1);
-          ub = Big*ones(nL+nP+(mC+2)*nC,1);
-          D = vertcat(D{:});
-          % just keep the likely contacts (and store data to check the unlikely):
-          phi_check = [phi_check;phiC(~possible_contact_indices)];
-          J_check = [J_check; n(~possible_contact_indices,:)];
-          phiC = phiC(possible_contact_indices);
-          n = n(possible_contact_indices,:);
-          D = D(repmat(possible_contact_indices,mC,1),:);
-          mu = mu(possible_contact_indices,:);
-
-          if isempty(obj.LCP_cache.data.possible_contact_indices) || any(obj.LCP_cache.data.possible_contact_indices~=possible_contact_indices)
-              possible_indices_changed = true;
-          end
-          obj.LCP_cache.data.possible_contact_indices=possible_contact_indices;
-
-          J(nL+nP+(1:nC),:) = n;
-          J(nL+nP+nC+(1:mC*nC),:) = D;
-
-          if nargout>4
-            dJ = zeros(nL+nP+(mC+2)*nC,num_q^2);  % was sparse, but reshape trick for the transpose below didn't work
-            dJ(nL+nP+(1:nC),:) = reshape(dn(possible_contact_indices,:),nC,[]);
-            dD = cellfun(@(A)reshape(A(possible_contact_indices,:),size(D{1},1),size(D{1},2)*size(dD{1},2)),dD,'UniformOutput',false);
-            dD = vertcat(dD{:});
-            dJ(nL+nP+nC+(1:mC*nC),:) = dD;
-          end
-
-          contact_data.normal = normal(:,possible_contact_indices);
-          for i=1:length(d)
-            contact_data.d{i} = d{i}(:,possible_contact_indices);
+          if numel(phiC > 0)
```

Comment:
ah. that makes sense.  but then you have a typo. sounds like you want `numel(phiC)>0` not  `numel(phiC>0)` ?

is `numel` faster than `isempty`?  (which is what we use almost everywhere else)

On Apr 23, 2015, at 3:22 PM, John Carter <notifications@github.com> wrote:

> In systems/plants/TimeSteppingRigidBodyManipulator.m:
> 
> > -
> > -          J(nL+nP+(1:nC),:) = n;
> > -          J(nL+nP+nC+(1:mC*nC),:) = D;
> > -
> > -          if nargout>4
> > -            dJ = zeros(nL+nP+(mC+2)*nC,num_q^2);  % was sparse, but reshape trick for the transpose below didn't work
> > -            dJ(nL+nP+(1:nC),:) = reshape(dn(possible_contact_indices,:),nC,[]);
> > -            dD = cellfun(@(A)reshape(A(possible_contact_indices,:),size(D{1},1),size(D{1},2)*size(dD{1},2)),dD,'UniformOutput',false);
> > -            dD = vertcat(dD{:});
> > -            dJ(nL+nP+nC+(1:mC*nC),:) = dD;
> > -          end
> > -
> > -          contact_data.normal = normal(:,possible_contact_indices);
> > -          for i=1:length(d)
> > -            contact_data.d{i} = d{i}(:,possible_contact_indices);
> > +          if numel(phiC > 0)
> phiC is the vector of the closest distances between actual contact pairs. If the vector is empty, there are no contact pairs.
> 
> getNumContactPairs returns a static value based on the number of possible contact pairs but bullet may decide at runtime that none of those pairs are relevant.
> 
> If there are no possible collision pairs, we can skip the call to collisionDetect but collisionDetect should have the ultimate say in whether or not collision pairs should actually be processed.
> 
> —
> Reply to this email directly or view it on GitHub.
>